src/librbd/ImageState.cc

```diff
@@ -229,18 +230,191 @@ class ImageUpdateWatchers {
+    notify(UNQUIESCE, on_finish);
+  }
+
+private:
+  enum EventType {QUIESCE, UNQUIESCE};
+
+  class ThreadPoolSingleton : public ThreadPool {
+  public:
+    explicit ThreadPoolSingleton(CephContext *cct)
+      : ThreadPool(cct, "librbd::QuiesceWatchers::thread_pool",
+                   "tp_librbd", 1) {
+      start();
+    }
+    ~ThreadPoolSingleton() override {
+      stop();
+    }
+  };
+
+  CephContext *m_cct;
+  ceph::mutex m_lock;
+  ContextWQ *m_work_queue = nullptr;
+  std::map<uint64_t, QuiesceWatchCtx*> m_watchers;
+  uint64_t m_next_handle = 0;
+  Context *m_on_notify = nullptr;
+  std::map<uint64_t, Context*> m_pending_unregister;
+
+  void notify(EventType event_type, Context *on_finish) {
+    ldout(m_cct, 20) << "QuiesceWatchers::" << __func__ << " " << event_type
+                     << dendl;
+
+    auto ctx = create_context_callback<
+      QuiesceWatchers, &QuiesceWatchers::handle_notify>(this);
+    auto gather_ctx = new C_Gather(m_cct, ctx);
+    {
+      std::lock_guard locker{m_lock};
+
+      ceph_assert(m_on_notify == nullptr);
+      m_on_notify = on_finish;
+
+      for (auto it : m_watchers) {
+        send_notify(it.first, it.second, event_type, gather_ctx->new_sub());
+      }
+    }
+
+    gather_ctx->activate();
+  }
+
+  void send_notify(uint64_t handle, QuiesceWatchCtx *watcher,
+                   EventType event_type, Context *on_finish) {
+    auto ctx = new LambdaContext(
+      [this, handle, watcher, event_type, on_finish](int) {
+        ldout(m_cct, 20) << "QuiesceWatchers::" << __func__ << ": handle="
+                         << handle << ", event_type=" << event_type << dendl;
+        switch (event_type) {
+        case QUIESCE:
+          watcher->handle_quiesce();
+          break;
+        case UNQUIESCE:
+          watcher->handle_unquiesce();
+          break;
+        default:
+          ceph_abort_msgf("invalid event_type %d", event_type);
+        }
+
+        on_finish->complete(0);
+      });
+
+    m_work_queue->queue(ctx);
+  }
+
+  void handle_notify(int r) {
+    ldout(m_cct, 20) << "QuiesceWatchers::" << __func__ << ": r=" << r
+                     << dendl;
+
+    ceph_assert(r == 0);
+
+    Context *on_notify = nullptr;
+    std::map<uint64_t, Context*> pending_unregister;
+    {
+      std::lock_guard locker{m_lock};
+
+      ceph_assert(m_on_notify != nullptr);
+      std::swap(on_notify, m_on_notify);
+      std::swap(pending_unregister, m_pending_unregister);
+    }
+
+    on_notify->complete(0);
+
+    for (auto &it : pending_unregister) {
+      ldout(m_cct, 20) << "QuiesceWatchers::" << __func__
+                       << ": completing unregister " << it.first << dendl;
+      it.second->complete(0);
+    }
+  }
+
+  void create_work_queue() {
+    if (m_work_queue != nullptr) {
+      return;
+    }
+    auto& thread_pool = m_cct->lookup_or_create_singleton_object<
+      ThreadPoolSingleton>("librbd::QuiesceWatchers::thread_pool",
```

Comment:
BTW, it reminds me about the question I wanted to ask and forgot!

The current behavior, with the default 1 op thead, is that if several callbacks are registered they are executed sequentially, which may increase total execution callback time causing unnecessary  timeouts. Not sure if we want to change/improve this somehow (have a thread per callback?)  though -- in real life I doubt someone would want to register several callback.

Reply:
I could see `rbd-nbd`'s future state registering one callback per image -- but it would have multiple librbd Images open. I would still think `rbd-nbd` should be responsible for executing the  freeze out-of-band and then invoke another "quiesce complete" API method when finished and ready. 

We also might need to handle reference counting on the pending quiesce / unquiesce callbacks in case two clients attempt to create a snapshot concurrently.